<a href="https://colab.research.google.com/github/brandonrecoder8/classwork2026/blob/main/data%20/load_explore.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Use the loaded housing_df
df = housing_df.copy() # Make a copy to avoid modifying the original loaded df

# Features and target
# Selecting relevant features based on initial inspection
numerical_features = ['sqft_living', 'bedrooms', 'bathrooms', 'floors', 'grade']
categorical_features = ['zipcode'] # Treating zipcode as categorical

X = df[numerical_features + categorical_features]
y = df['price']

# Preprocessing: Scale numerical features and one-hot encode categorical features
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numerical_features),
        ('cat', OneHotEncoder(handle_unknown='ignore', sparse_output=False), categorical_features)
    ], remainder='passthrough')

# Create pipeline with preprocessing and model
model = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('regressor', LinearRegression())
])

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2,
                                                    random_state=42)

# Train model
model.fit(X_train, y_train)

# Evaluate the model
r_squared = model.score(X_test, y_test)
print(f"Model R-squared on the test set: {r_squared:.4f}")

# Display model coefficients (simplified)
# feature_names_processed = []
# num_features_out = numerical_features
# cat_features_out = list(model.named_steps['preprocessor'].named_transformers_['cat'].get_feature_names_out(categorical_features))
# feature_names_processed = num_features_out + cat_features_out

# coefficients = model.named_steps['regressor'].coef_

# print("\nModel Coefficients (first 10, for illustration):")
# for feature, coef in zip(feature_names_processed[:10], coefficients[:10]):
#     print(f"{feature}: {coef:.2f}")

Model R-squared on the test set: 0.7478


In [ ]:
housing_df = pd.read_csv('/content/Housing.csv')
print("Housing Data Head:")
display(housing_df.head())
print("\nHousing Data Info:")
housing_df.info()

Housing Data Head:


,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
0,7229300521,20141013T000000,231300.0,2,1.00,1180,5650,1.0,0,0,...,7,1180,0,1955,0,98178,47.5112,-122.257,1340,5650
1,6414100192,20141209T000000,538000.0,3,2.25,2570,7242,2.0,0,0,...,7,2170,400,1951,1991,98125,47.7210,-122.319,1690,7639
2,5631500400,20150225T000000,180000.0,2,1.00,770,10000,1.0,0,0,...,6,770,0,1933,0,98028,47.7379,-122.233,2720,8062
3,2487200875,20141209T000000,604000.0,4,3.00,1960,5000,1.0,0,0,...,7,1050,910,1965,0,98136,47.5208,-122.393,1360,5000
4,1954400510,20150218T000000,510000.0,3,2.00,1680,8080,1.0,0,0,...,8,1680,0,1987,0,98074,47.6168,-122.045,1800,7503



Housing Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null 

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score

# Use the loaded telecom_churn_df
df = telecom_churn_df.copy() # Make a copy to avoid modifying the original loaded df

# Features and target
# All columns except 'Churn' are features
feature_columns = [col for col in df.columns if col != 'Churn']
numerical_features = feature_columns # All features appear to be numerical

X = df[numerical_features]
y = df['Churn']

# Preprocessing: Scale numerical features
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numerical_features)
    ], remainder='passthrough')

# Create pipeline with preprocessing and model
model = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', LogisticRegression(random_state=42))
])

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2,
                                                    random_state=42)

# Train model
model.fit(X_train, y_train)

# Evaluate the model
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Model Accuracy on the test set: {accuracy:.4f}")

# Coefficients are less directly interpretable after scaling and can be many with one-hot encoding,
# so we'll omit them for brevity here. If needed, they can be extracted and mapped back.
# feature_names = (model.named_steps['preprocessor'].named_transformers_['cat'].get_feature_names_out(['region'])).tolist() + ['age','monthly_usage_hours', 'purchase_amount', 'customer_service_calls']
# coefficients = model.named_steps['classifier'].coef_[0]
# print("\nModel Coefficients:")
# for feature, coef in zip(feature_names, coefficients):
#     print(f"{feature}: {coef:.2f}")

Model Accuracy on the test set: 0.8591


In [ ]:
telecom_churn_df = pd.read_csv('/content/telecom_churn.csv')
print("Telecom Churn Data Head:")
display(telecom_churn_df.head())
print("\nTelecom Churn Data Info:")
telecom_churn_df.info()

Telecom Churn Data Head:


,Churn,AccountWeeks,ContractRenewal,DataPlan,DataUsage,CustServCalls,DayMins,DayCalls,MonthlyCharge,OverageFee,RoamMins
0,0,128,1,1,2.7,1,265.1,110,89.0,9.87,10.0
1,0,107,1,1,3.7,1,161.6,123,82.0,9.78,13.7
2,0,137,1,0,0.0,0,243.4,114,52.0,6.06,12.2
3,0,84,0,0,0.0,2,299.4,71,57.0,3.10,6.6
4,0,75,0,0,0.0,3,166.7,113,41.0,7.42,10.1



Telecom Churn Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Churn            3333 non-null   int64  
 1   AccountWeeks     3333 non-null   int64  
 2   ContractRenewal  3333 non-null   int64  
 3   DataPlan         3333 non-null   int64  
 4   DataUsage        3333 non-null   float64
 5   CustServCalls    3333 non-null   int64  
 6   DayMins          3333 non-null   float64
 7   DayCalls         3333 non-null   int64  
 8   MonthlyCharge    3333 non-null   float64
 9   OverageFee       3333 non-null   float64
 10  RoamMins         3333 non-null   float64
dtypes: float64(5), int64(6)
memory usage: 286.6 KB


In [ ]:
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt

# Use the loaded customer_segmentation_df
df = customer_segmentation_df.copy() # Make a copy to avoid modifying the original loaded df

# Data Cleaning and Feature Engineering
# Drop rows with missing 'Income' for simplicity, or impute. Let's impute with median for robustness.
df['Income'] = df['Income'].fillna(df['Income'].median())

# Create Age feature
current_year = pd.Timestamp.now().year
df['Age'] = current_year - df['Year_Birth']

# Select relevant numerical features for clustering
# Based on typical customer segmentation analyses and available data fields
clustering_features = [
    'Income',
    'Age',
    'Recency',
    'MntWines', 'MntFruits', 'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts', 'MntGoldProds',
    'NumDealsPurchases', 'NumWebPurchases', 'NumCatalogPurchases', 'NumStorePurchases',
    'NumWebVisitsMonth',
    'Kidhome', 'Teenhome'
]

X = df[clustering_features]

# Preprocess data: Scale numerical features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Determine optimal number of clusters using elbow method
inertia = []
K = range(1, 11) # Test up to 10 clusters
for k in K:
    kmeans = KMeans(n_clusters=k, random_state=42, n_init=10) # Set n_init to suppress warning
    kmeans.fit(X_scaled)
    inertia.append(kmeans.inertia_)

# Plot elbow curve
plt.figure(figsize=(10, 6))
plt.plot(K, inertia, 'bo-')
plt.xlabel('Number of Clusters (K)')
plt.ylabel('Inertia')
plt.title('Elbow Method for Optimal K')
plt.grid(True)
plt.savefig('elbow_plot_segmentation.png') # Save with a distinct name
plt.close()

print("Elbow plot saved as elbow_plot_segmentation.png. Please examine it to choose an optimal K.")

# For demonstration, let's assume optimal_k = 4 (this needs to be determined by looking at the plot)
# User would typically review the elbow plot and decide.
# For now, we will pick a reasonable number for demonstration.
optimal_k = 4
print(f"\nProceeding with optimal_k = {optimal_k} for K-Means clustering (based on a typical elbow point evaluation).")

kmeans = KMeans(n_clusters=optimal_k, random_state=42, n_init=10)
df['cluster'] = kmeans.fit_predict(X_scaled)

# Analyze clusters (using original features for better interpretability)
cluster_summary = df.groupby('cluster')[clustering_features].mean().round(2)
print("\nCluster Characteristics:")
print(cluster_summary)

# Example of targeted strategies (adapt based on new features)
print("\nExample of Targeted Strategies based on new clusters:")
for cluster in range(optimal_k):
    print(f"\nCluster {cluster} Strategy:")
    current_cluster_summary = cluster_summary.loc[cluster]
    if current_cluster_summary['Income'] > df['Income'].median() and current_cluster_summary['MntWines'] > df['MntWines'].median():
        print("High-income, high-spending (wine) customers: Offer premium wine club memberships or exclusive events.")
    elif current_cluster_summary['NumDealsPurchases'] > df['NumDealsPurchases'].median() and current_cluster_summary['Recency'] < df['Recency'].median():
        print("Deal-seeking, recent buyers: Promote new deals and bundle offers.")
    elif current_cluster_summary['Kidhome'] > 0 or current_cluster_summary['Teenhome'] > 0:
        print("Family-focused customers: Highlight family-friendly products or discounts on larger quantities.")
    else:
        print("General customers: Engage with personalized recommendations based on past purchases.")

# Save cluster assignments to CSV
df[['ID', 'cluster'] + clustering_features].to_csv('customer_segments_full.csv', index=False)
print("\nCustomer segments with assigned clusters saved to customer_segments_full.csv")

Elbow plot saved as elbow_plot_segmentation.png. Please examine it to choose an optimal K.

Proceeding with optimal_k = 4 for K-Means clustering (based on a typical elbow point evaluation).

Cluster Characteristics:
           Income    Age  Recency  MntWines  MntFruits  MntMeatProducts  \
cluster                                                                   
0        29884.23  47.49    48.30     35.86       6.63            28.04   
1        76491.88  57.58    49.40    582.65      69.69           453.50   
2        59634.48  61.41    47.91    527.15      22.47           146.91   
3        42398.31  62.68    50.79     68.89       4.30            25.84   

         MntFishProducts  MntSweetProducts  MntGoldProds  NumDealsPurchases  \
cluster                                                                       
0                   9.87              6.75         18.43               1.93   
1                 100.23             71.60         76.97               1.31   
2                

In [ ]:
customer_segmentation_df = pd.read_csv('/content/customer_segmentation.csv')
print("Customer Segmentation Data Head:")
display(customer_segmentation_df.head())
print("\nCustomer Segmentation Data Info:")
customer_segmentation_df.info()

Customer Segmentation Data Head:


,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
0,5524,1957,Graduation,Single,58138.0,0,0,04-09-2012,58,635,...,7,0,0,0,0,0,0,3,11,1
1,2174,1954,Graduation,Single,46344.0,1,1,08-03-2014,38,11,...,5,0,0,0,0,0,0,3,11,0
2,4141,1965,Graduation,Together,71613.0,0,0,21-08-2013,26,426,...,4,0,0,0,0,0,0,3,11,0
3,6182,1984,Graduation,Together,26646.0,1,0,10-02-2014,26,11,...,6,0,0,0,0,0,0,3,11,0
4,5324,1981,PhD,Married,58293.0,1,0,19-01-2014,94,173,...,5,0,0,0,0,0,0,3,11,0



Customer Segmentation Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumD

Summary:

1. Housing Price Prediction Model
Original (Starter) Code: The initial code for the housing price prediction model used synthetic data with features like square_footage (numerical) and location (categorical), and price as the target. It performed basic preprocessing (one-hot encoding for location) and trained a LinearRegression model, then predicted a price for a new house and displayed coefficients.

New Dataset Integration (Housing.csv): We introduced the Housing.csv dataset, which contains a rich set of features related to house sales, including price, sqft_living, bedrooms, bathrooms, floors, grade, and zipcode, among others.

Changes Made and Justification:

Data Source: The df = pd.DataFrame(data) line for synthetic data was replaced with df = housing_df.copy() to utilize the actual Housing.csv data.
Feature Selection: Instead of ['square_footage', 'location'], we selected numerical_features = ['sqft_living', 'bedrooms', 'bathrooms', 'floors', 'grade'] and categorical_features = ['zipcode']. This was done to use more realistic and relevant features from the new dataset for predicting house prices.
Preprocessing Enhancement:
StandardScaler was added for numerical_features. This is crucial because real-world features like square footage and number of bathrooms often have different scales, which can impact the performance of models like Linear Regression. Scaling ensures all numerical features contribute equally to the model.
OneHotEncoder was applied to zipcode. zipcode is a categorical identifier, not a numerical value, so one-hot encoding converts it into a format that the model can understand without implying ordinal relationships.
The ColumnTransformer was updated to handle both numerical scaling and categorical encoding.
Model Evaluation: The code was updated to print the R-squared score on the test set (model.score(X_test, y_test)). This provides a standard metric to evaluate how well the model explains the variance in house prices, which is more appropriate than just printing a single prediction from sample data.
Prediction and Coefficient Display: The specific prediction for a synthetic new house and the display of coefficients were commented out as they were less relevant after switching to a full dataset and focusing on overall model performance metrics.
How the New Code Functions: The updated code now loads the real housing data, preprocesses it by scaling numerical features and one-hot encoding the zipcode, splits the data into training and testing sets, trains a LinearRegression model, and then evaluates its performance using the R-squared score on unseen test data. This provides a robust assessment of the model's predictive power on actual housing market data.

2. Customer Churn Prediction Model
Original (Starter) Code: The initial customer churn prediction model used synthetic data with features like age, monthly_usage_hours, purchase_amount, customer_service_calls (numerical), and region (categorical), with churn as the binary target. It performed scaling for numerical features and one-hot encoding for region and trained a LogisticRegression model, predicting churn probability for a sample customer.

New Dataset Integration (telecom_churn.csv): We integrated the telecom_churn.csv dataset, which contains various customer attributes and a 'Churn' column indicating whether a customer churned or not.

Changes Made and Justification:

Data Source: The synthetic data generation was replaced with df = telecom_churn_df.copy() to use the actual telecom churn data.
Feature Selection: The features were updated to feature_columns = [col for col in df.columns if col != 'Churn'], and numerical_features = feature_columns. This approach dynamically selects all columns except the target 'Churn' as features, as inspection showed them to be primarily numerical. The original 'region' categorical feature was not present in this dataset, simplifying the preprocessing.
Preprocessing Simplification: Since all selected features in telecom_churn.csv were numerical, the ColumnTransformer was simplified to only include StandardScaler for all numerical features. OneHotEncoder was removed as there were no categorical features requiring it in this specific dataset.
Model Evaluation: The code was updated to calculate and print the accuracy_score on the test set. For classification tasks like churn prediction, accuracy is a common and intuitive metric to gauge model performance, replacing the single sample prediction of churn probability.
Prediction and Coefficient Display: The specific prediction for a synthetic new customer and detailed coefficient display were commented out, again to focus on a broader model evaluation with the full dataset.
How the New Code Functions: The updated code now loads the telecom churn data, preprocesses all its numerical features using StandardScaler, splits the data, trains a LogisticRegression model, and then evaluates its performance using the accuracy score on the test set. This provides a more generalized and statistically sound evaluation of the churn prediction model.

3. Customer Segmentation using K-Means
Original (Starter) Code: The initial customer segmentation model used synthetic data with features like annual_spending, purchase_frequency, and age. It scaled these features, applied the elbow method to suggest an optimal K (number of clusters), and then performed KMeans clustering, saving a basic elbow_plot.png and customer_segments.csv.

New Dataset Integration (customer_segmentation.csv): We incorporated the customer_segmentation.csv dataset, which is much richer, containing demographic information (Year_Birth, Education, Marital_Status), spending habits across various product categories (MntWines, MntFruits, etc.), purchase channels (NumWebPurchases, NumStorePurchases), and campaign responses.

Changes Made and Justification:

Data Source: The synthetic data generation was replaced with df = customer_segmentation_df.copy() to use the real customer segmentation data.
Data Cleaning and Feature Engineering:
Missing Income: The Income column had missing values. We added df['Income'] = df['Income'].fillna(df['Income'].median()) to impute these missing values with the median income. This is a robust way to handle missing numerical data, preventing errors and retaining as much data as possible.
Age Derivation: A new Age feature was created from Year_Birth using current_year - df['Year_Birth']. Age is often a critical demographic factor for segmentation, and deriving it provides a more directly usable feature.
Expanded Clustering Features: The clustering_features list was significantly expanded to include a wide range of relevant numerical columns from the customer_segmentation.csv dataset, such as Income, Age, Recency, various Mnt (amount spent) columns, and Num (number of purchases) columns. This leverages the rich detail of the new dataset to create more meaningful customer segments.
Elbow Plot Naming: The elbow plot filename was changed to elbow_plot_segmentation.png to avoid overwriting or confusion if multiple elbow plots were generated.
Optimal K Assumption: A comment was added, clarifying that optimal_k = 4 is an assumption for demonstration purposes, and users should review the generated elbow plot to make an informed decision about the true optimal K.
Enhanced Cluster Analysis and Strategies: The example of targeted strategies was significantly expanded and made conditional based on characteristics of the new features (e.g., Income, MntWines, NumDealsPurchases, Kidhome, Teenhome). This demonstrates how the richer insights from the new dataset can lead to more nuanced and actionable marketing strategies for each segment.
Output CSV: The saved CSV filename was changed to customer_segments_full.csv, and it now explicitly includes the ID and clustering_features along with the cluster assignment, providing a more complete output for further analysis.
How the New Code Functions: The updated code first loads the comprehensive customer segmentation data. It performs essential data cleaning (imputing missing income) and feature engineering (calculating age). It then selects a broad set of features relevant to customer behavior, scales them, and applies the KMeans algorithm. The elbow method is still used to aid in choosing K. Finally, it provides a detailed summary of each cluster's characteristics based on the original (unscaled) features and suggests targeted marketing strategies tailored to these richer segments, saving the results to a new CSV file. This allows for a much more insightful and actionable customer segmentation.

Detailed description of changes made:

Housing Price Prediction Model (cell UE7uItUgWM8z): The model was updated with Housing.csv, using sqft_living, bedrooms, bathrooms, floors, grade, and one-hot encoded zipcode as features. It achieved an R-squared score of 0.7478 on the test set, indicating a reasonably good fit for predicting house prices.

Customer Churn Prediction Model (cell x5T_pqQgWi6B): This model was updated with telecom_churn.csv, using all non-'Churn' columns as numerical features after scaling. It achieved an accuracy score of 0.8591 on the test set, demonstrating good performance in predicting customer churn.

Customer Segmentation using K-Means (cell sUZ1gfs-WwNC): The segmentation model was updated with customer_segmentation.csv. Missing 'Income' values were imputed, 'Age' was derived, and several features related to spending and demographics were used for clustering. An elbow plot (elbow_plot_segmentation.png) was generated to help determine the optimal number of clusters. We proceeded with optimal_k = 4 to analyze the cluster characteristics and generate targeted strategies. The customer segments were saved to customer_segments_full.csv